src/main.rs

```diff
@@ -649,29 +650,69 @@ impl App {
     }
 
     fn document_statistics(&self) -> Element<Message> {
-        //TODO: calculate in the background
-        let mut character_count = 0;
-        let mut character_count_no_spaces = 0;
+        let (sender, receiver) = crossbeam::channel::unbounded();
         let mut line_count = 0;
+        let mut  lines: Vec<BufferLine> = Vec::new();
+
+        fn count_words(text: &str) -> usize {
+            let mut in_word = false;
+            let mut word_count = 0;
+
+            for c in text.chars() {
+                if !c.is_whitespace() {
+                    in_word = true;
+                } else if in_word {
+                    word_count += 1;
+                    in_word = false;
+                }
+            }
+
+            // Handle the final word if it doesn't end with whitespace
+            if in_word {
+                word_count += 1;
+            }
+
+            word_count
+        }
+
+
+
+        fn count_characters_and_words(lines : &Vec<BufferLine>) -> (usize, usize, usize) {
+
+            let mut character_count = 0;
+            let mut character_count_no_spaces = 0;
+            let mut words_count = 0;
+
+            for line in  lines.iter() {
+                words_count += count_words(line.text());
+                for c in line.text().chars() {
+                    character_count += 1;
+                    if !c.is_whitespace() {
+                        character_count_no_spaces += 1;
+                    }
+                }
+            }
+
+            (character_count, character_count_no_spaces, words_count)
+        }
+
+
 
         if let Some(Tab::Editor(tab)) = self.active_tab() {
             let editor = tab.editor.lock().unwrap();
             editor.with_buffer(|buffer| {
                 line_count = buffer.lines.len();
-                for line in buffer.lines.iter() {
-                    //TODO: do graphemes?
-                    for c in line.text().chars() {
-                        character_count += 1;
-                        if !c.is_whitespace() {
-                            character_count_no_spaces += 1;
-                        }
-                    }
-                }
+                lines = buffer.lines.clone();
+                thread::spawn(move || sender.send(count_characters_and_words(&lines)).unwrap());
+
+
             });
         }
 
+        let  (character_count,character_count_no_spaces,words_count) = receiver.recv().unwrap();
```

Comment:
This doesn't really run it in the background if the gui has to block for it to finish. Look at the Command::perform function to see how this can be done.

Reply:
okay